### PR TITLE
Marc gine

### PR DESCRIPTION
**Pull request by ginesabata**
The 'Train' function has been added "without pretraining".
    There are black-box optimizer and function components, which, if necessary, could be programmed in "utils", although they 
     may not be as efficient.

Hyperparameters used in the training have been added.

A precision function has been added to "utils". This point is based on practice 3 of the subject "Projects".

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    # First trials, just one class: Line road??
+    model = Model1(num_classes=1).to(device)  # Just one class: "LINE ROAD"?<-
+    # Initialize optimizer (possibility to create specific optimizer call from utils)
+    # Self-drive examples found use the Adam optimizer
+    optimizer = optim.Adam(model.parameters(), lr=hparams['learning_rate'],weight_decay=hparams['weight_decay']) #Note use weight_decay to prevent overfitting
+
+    for epoch in range(hparams['num_epochs']):  
+        # Train model for 1 epoch
+        train_loss, train_acc = train_single_epoch(model, train_loader, optimizer)
+        print(f"Train Epoch {epoch} loss={train_loss:.2f} acc={train_acc:.2f}")
```
Comment (MarcRamonMoreno) [inline src/train.py:57]: Aqui en comptes de printf podem enviar aquest missatge per logging.info
`        logging.info(f"Train Epoch {epoch} loss={train_loss:.2f} acc={train_acc:.2f}")`

File: src/train.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    return model
```
Comment (MarcRamonMoreno) [inline src/train.py:58]: Abans de retornar el model pots guardar el millor model i la millor train_loss i enviar-la per logging.info.
```python
        # Save the best model
        if train_loss < best_loss:
            best_loss = train_loss
            torch.save(model.state_dict(), 'best_model.pth') # S'hauria d'especificar el path per guardar aquest model
            logging.info(f"Saved best model with loss {best_loss:.2f}")
```

File: src/train.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    # Add loss and accuracy mean
+    return np.mean(losses), np.mean(accs)
+
+def train_model(hparams):
+    # Generate Training dataset
+    train_dataset = MyDataset("train_images_path", "train_masks_path")  # ->REQUIRED UPDATE PATH<-
+    # Generate DataLoader
+    train_loader = DataLoader(train_dataset, batch_size=hparams["batch_size"], shuffle=True)
+    # First trials, just one class: Line road??
+    model = Model1(num_classes=1).to(device)  # Just one class: "LINE ROAD"?<-
```
[the remark this answers is not in the document]
Comment (jsabahu) [reply src/train.py:49]: Segurament seran dues. 1 pel background, y laltre per les lines del carril

**Review by jsabahu — CHANGES_REQUESTED** [review]
